### PR TITLE
hier_block2: prevent crash/freeze after disconnection

### PR DESCRIPTION
**Pull request by vladisslav2011**
Signed-off-by: Vladisslav P <vladisslav@inbox.ru>

# Pull Request Details
Fix freeze on hier_block2::lock() when hier_block2 is just created.
Fix freeze on hier_block2::lock() when was disconnected from another block
and disconnect(hier_block...) is the last operation before  hier_block2::lock()

## Description
1. Check for top_block correctly with dynamic_cast
2. Use weak_ptr to track parent block
3. Maintain reference counter and clear d_parent when all ports get disconnected
4. Clear d_parent on disconnect_all
5. Add tests
6. Limit test run time to 30 seconds to make it fail in predictable time (blocks only)

## Related Issue
Close https://github.com/gnuradio/gnuradio/issues/5436

## Which blocks/areas does this affect?
hier_block2

## Testing Done
C++ test `main.cpp`:
```
#include <iostream>

#include <gnuradio/blocks/null_sink.h>
#include <gnuradio/blocks/null_source.h>
#include <gnuradio/hier_block2.h>
#include <gnuradio/blocks/copy.h>
#include <gnuradio/top_block.h>


class test_block : public gr::hier_block2
{

public:
typedef std::shared_ptr<test_block> sptr;
    static test_block::sptr make()
    {
        return gnuradio::get_initial_sptr(new test_block());
    }
    test_block();
    virtual ~test_block();
    void trigger();

private:
    gr::blocks::copy::sptr d_copy;
};

test_block::test_block(): gr::hier_block2 ("test_block",
                          gr::io_signature::make (1, 1, sizeof(gr_complex)),
                          gr::io_signature::make (1, 1, sizeof(gr_complex)))

{
    d_copy = gr::blocks::copy::make(sizeof(gr_complex));
    connect(self() , 0 , d_copy , 0);
    connect(d_copy , 0, self() , 0);
}

test_block::~test_block()
{
}

void test_block::trigger()
{
    std::cerr<<"test_block->trigger"<<std::endl;
    lock();
    std::cerr<<"test_block->lock"<<std::endl;
    unlock();
    std::cerr<<"test_block->unlock"<<std::endl;
}

int main()
{
    auto tb = gr::make_top_block("buggy");
    auto test0 = test_block::make();
    auto src0 = gr::blocks::null_source::make(sizeof(gr_complex));
    auto snk0 = gr::blocks::null_sink::make(sizeof(gr_complex));
    auto snk1 = gr::blocks::null_sink::make(sizeof(gr_complex));
    test0->lock();
    test0->unlock();
    std::cerr<<"Unconnected lock test successful."<<std::endl;
    tb->lock();
    std::cerr<<"tb->lock"<<std::endl;
    tb->connect(src0, 0, test0, 0);
    tb->connect(test0, 0, snk0, 0);
    tb->connect(test0, 0, snk1, 0);
    tb->unlock();
    std::cerr<<"tb->unlock"<<std::endl;
    tb->start();
    std::cerr<<"tb->start"<<std::endl;
    tb->lock();
    std::cerr<<"tb->lock"<<std::endl;
    tb->disconnect(test0, 0, snk1, 0);
    tb->unlock();
    std::cerr<<"tb->unlock"<<std::endl;
    test0->trigger();
    std::cerr<<"TEST PASSED!!!"<<std::endl;
    tb->lock();
    tb->disconnect_all();
    std::cerr<<"tb->disconnect_all"<<std::endl;
    tb->connect(src0, 0, snk0, 0);
    tb->unlock();
    std::cerr<<"tb->unlock"<<std::endl;
    test0->trigger();
    std::cerr<<"TEST 2PASSED!!!"<<std::endl;
    return 0;
}

```
Compile: `g++ main.cpp -o main -lboost_system -lfmt -lgnuradio-runtime -lgnuradio-pmt -lvolk -lgnuradio-blocks`
Run: `./main`
Expected output:
```
Unconnected lock test successful.
tb->lock
tb->unlock
tb->start
tb->lock
tb->unlock
test_block->trigger
test_block->lock
test_block->unlock
TEST PASSED!!!
tb->disconnect_all
tb->unlock
test_block->trigger
test_block->lock
test_block->unlock
TEST 2PASSED!!!
```

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/master/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [x] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

File: gnuradio-runtime/lib/hier_block2_detail.cc [hunk 1]
```diff
@@ -927,5 +937,77 @@ std::string hier_block2_detail::log_level()
...
+    if (old_parent.get() == parent) {
+        d_parent_refcnt++;
+        return;
+    }
+    std::stringstream msg;
+    msg << "Hierarchical blocks do not yet support moving from one hierarchy "
+           "to another. Block \""
+        << d_owner->name() << "\" moved from \"" << old_parent->name() << "\" to \""
+        << parent->name() << "\"";
+    throw std::runtime_error(msg.str());
```
Comment (argilo) [inline gnuradio-runtime/lib/hier_block2_detail.cc:957]: Throwing here looks like it could break existing applications.

**Comment by argilo** [comment]
I think it would be worth adding some regression tests to the CI test suite.

The test code posted in the pull request description here, as well as in #5436, could be adapted for that purpose.

**Comment by argilo** [comment]
For reference, here are some of the current tests involving `lock()` and `unlock()`, which could also be used as a model:

https://github.com/gnuradio/gnuradio/blob/3d4d9a4245b9e126d7119a7c81905c0d5423233b/gr-blocks/python/blocks/qa_hier_block2.py#L324-L335

https://github.com/gnuradio/gnuradio/blob/3d4d9a4245b9e126d7119a7c81905c0d5423233b/gr-blocks/python/blocks/qa_hier_block2.py#L440-L477

https://github.com/gnuradio/gnuradio/blob/3d4d9a4245b9e126d7119a7c81905c0d5423233b/gr-blocks/lib/qa_gr_top_block.cc#L70-L115

File: gnuradio-runtime/lib/hier_block2_detail.cc (hunk 1 above)
Comment (vladisslav2011) [reply gnuradio-runtime/lib/hier_block2_detail.cc:957]: Throwing may occur only if some block is connected inside one hier_block2 and then connected inside another hier_block2 without being fully disconnected from first hier_block2. I don't think this to be correct usage of hier_block2.
Exception test:
```
#include <iostream>

#include <gnuradio/blocks/null_sink.h>
#include <gnuradio/blocks/null_source.h>
#include <gnuradio/hier_block2.h>
#include <gnuradio/blocks/copy.h>
#include <gnuradio/top_block.h>


class test_block : public gr::hier_block2
{

public:
typedef std::shared_ptr<test_block> sptr;
    static test_block::sptr make()
    {
        return gnuradio::get_initial_sptr(new test_block());
    }
    test_block();
    virtual ~test_block();

    private:
    gr::blocks::copy::sptr d_copy;
};

test_block::test_block(): gr::hier_block2 ("test_block",
                          gr::io_signature::make (1, 1, sizeof(gr_complex)),
                          gr::io_signature::make (1, 1, sizeof(gr_complex)))

{
    d_copy = gr::blocks::copy::make(sizeof(gr_complex));
    connect(self() , 0 , d_copy , 0);
    connect(d_copy , 0, self() , 0);
}

test_block::~test_block()
{
}

class test_block2 : public gr::hier_block2
{

public:
typedef std::shared_ptr<test_block2> sptr;
    static test_block2::sptr make()
    {
        return gnuradio::get_initial_sptr(new test_block2());
    }
    test_block2();
    virtual ~test_block2();

    test_block::sptr d_test;
};

test_block2::test_block2(): gr::hier_block2 ("test_block",
                          gr::io_signature::make (1, 1, sizeof(gr_complex)),
                          gr::io_signature::make (1, 1, sizeof(gr_complex)))

{
    d_test = test_block::make();
    connect(self() , 0 , d_test , 0);
    connect(d_test , 0, self() , 0);
}

test_block2::~test_block2()
{
}

int main()
{
    auto tb = gr::make_top_block("buggy");
    auto test0 = test_block2::make();
    auto test1 = test_block2::make();
    auto src0 = gr::blocks::null_source::make(sizeof(gr_complex));
    auto snk0 = gr::blocks::null_sink::make(sizeof(gr_complex));
    auto snk1 = gr::blocks::null_sink::make(sizeof(gr_complex));
    auto bad_copy = gr::blocks::copy::make(sizeof(gr_complex));
    tb->connect(src0, 0, test0, 0);
    tb->connect(src0, 0, test1, 0);
    tb->connect(test0, 0, snk0, 0);
    tb->connect(test1, 0, snk1, 0);
    test0->disconnect(test0->d_test, 0, test0->self(), 0);
    test1->disconnect(test1->d_test, 0, test1->self(), 0);
    test0->connect(test1->d_test, 0, test0->self(), 0);
    test1->connect(test0->d_test, 0, test1->self(), 0);
    tb->start();
    return 0;
}
```

**Comment by vladisslav2011** [comment]
> I think it would be worth adding some regression tests to the CI test suite.

On my machine, that builds gqrx in 7-10 seconds (`make clean ; make`) the test may run for several minutes without crashing in case of this bug being triggered.
Github CI builds gqrx in ~4 minutes, so it may hang for an hour before crash. Is it acceptable?
I think, the test should be implemented multi-threaded form with one thread supervising another: timed wait on condition variable, wait failed, crash with exception...
My attempt to run `make test` from current master result in freeze and python consuming 40Gb of RAM. After that I have not attempted to run `make test` again.

**Comment by argilo** [comment]
> the test may run for several minutes without crashing in case of this bug being triggered.

> I think, the test should be implemented in multi-threaded form with one thread supervising another: timed wait on condition variable, wait failed, crash with exception...

I don't think this complexity is necessary. As long as the tests run quickly when they pass, they should be fine. The tests will only fail if someone makes a change that breaks this functionality, and in that case I think it's OK to wait for the CI timeout to kick in.

File: gnuradio-runtime/lib/hier_block2_detail.cc (hunk 1 above)
Comment (argilo) [reply gnuradio-runtime/lib/hier_block2_detail.cc:957]: Got it. I think you're correct that an instance of a hierarchical block should only be present inside one parent hierarchical block at a time. I was confused by the error message though:

> Hierarchical blocks do not yet support moving from one hierarchy to another.

This makes it sound like you're not allowed to remove a hierarchical block from one parent and then add it to another. It also makes it sound like this feature will be allowed in the future. How about something like:

> A hierarchical block cannot have multiple parents. Block X must be removed from parent Y before being added to parent Z.

Comment (vladisslav2011) [reply gnuradio-runtime/lib/hier_block2_detail.cc:957]: I agree. The message is left here from the first version, that had no reference counting implemented. I'll update the message and add 2 tests. Lock on unconnected hier_block2 and lock after disconnection.

**Comment by vladisslav2011** [comment]
> I don't think this complexity is necessary.

Limiting test run time works well.
Tests ready.

**Comment by vladisslav2011** [comment]
Fixed formatting.

**Comment by vladisslav2011** [comment]
CI failure:
```
144/252 Test #144: qa_pfb_channelizer ...........................***Failed    0.55 sec
.F..
======================================================================
FAIL: test_0001 (__main__.test_pfb_channelizer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "$SRC_DIR/gr-filter/python/filter/qa_pfb_channelizer.py", line 53, in test_0001
    self.check_channelizer(filter.pfb.channelizer_hier_ccf(
  File "$SRC_DIR/gr-filter/python/filter/qa_pfb_channelizer.py", line 121, in check_channelizer
    self.compare_data(expected, received)
  File "$SRC_DIR/gr-filter/python/filter/qa_pfb_channelizer.py", line 129, in compare_data
    self.assertComplexTuplesAlmostEqual(expected, received, 3)
  File "$SRC_DIR/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 73, in assertComplexTuplesAlmostEqual
    return all([
  File "$SRC_DIR/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 74, in <listcomp>
    self.assertComplexAlmostEqual(x, y, places, msg)
  File "$SRC_DIR/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 37, in assertComplexAlmostEqual
    raise self.failureException(
AssertionError: (0.8881364488135434-0.45957986062149014j) != (nan+nanj) within 3 places

----------------------------------------------------------------------
Ran 4 tests in 0.344s

FAILED (failures=1)
```
It looks completely unrelated to changes applied. Math error...
This test is not failing on my machine:
```
python3 ../gr-filter/python/filter/qa_pfb_channelizer.py               
....
----------------------------------------------------------------------
Ran 4 tests in 0.271s

OK
```

**Comment by vladisslav2011** [comment]
Let's add python tests and try to rerun CI.

**Comment by vladisslav2011** [comment]
Python formatting fixed.

**Comment by willcode** [comment]
@vladisslav2011 The PFB channelizer is one of a few tests that fail randomly. Don't worry about it.

**Comment by mormj** [comment]
Fixes the test case given by @argilo in #5436 
@argilo - does this fix the issue as worked around in GQRX

**Review by mormj — APPROVED** [review]

**Comment by argilo** [comment]
@mormj I've got some free time tonight, so I'll take a closer look at this PR then and make sure that the code changes & tests address the issues that Gqrx was running into.

File: gr-blocks/python/blocks/qa_hier_block2.py [hunk 2]
```diff
@@ -476,6 +476,32 @@ def thread_01(hblock, cls):
...
+        hb.connect(m1, hb)
+        dst1 = blocks.vector_sink_f()
+        dst2 = blocks.vector_sink_f()
+        tb.connect(src, hb, dst1)
+        tb.connect(hb, dst2)
+        tb.run()
+        tb.disconnect(hb, dst2)
+        hb.lock()
+        hb.unlock()
+        tb.run()
```
Comment (argilo) [inline gr-blocks/python/blocks/qa_hier_block2.py:496]: This seems like the wrong order. A running flowgraph should be locked before it is modified, and unlocking it starts it running again. Also, it's sensible to wait for execution to finish before ending the test.
```suggestion
        tb.run()
        hb.lock()
        tb.disconnect(hb, dst2)
        hb.unlock()
        tb.wait()
```

Comment (argilo) [reply gr-blocks/python/blocks/qa_hier_block2.py:496]: I've verified that the test still works properly after those changes; it segfaults on master and passes on this branch.

File: gr-blocks/lib/qa_gr_hier_block2_derived.cc [hunk 3]
```diff
@@ -72,3 +72,33 @@ BOOST_AUTO_TEST_CASE(test_1)
...
+    tb->connect(head, 0, blk, 0);
+    tb->connect(blk, 0, dst0, 0);
+    tb->connect(blk, 0, dst1, 0);
+
+    tb->run();
+
+    tb->disconnect(blk, 0, dst1, 0);
+    blk->lock();
+    blk->unlock();
+    tb->run();
```
Comment (argilo) [inline gr-blocks/lib/qa_gr_hier_block2_derived.cc:103]: As with the Python test, I think the order should be changed here:
```suggestion
    blk->lock();
    tb->disconnect(blk, 0, dst1, 0);
    blk->unlock();
    tb->wait();
```
The test still crashes on master & passes on this branch.

File: gr-blocks/lib/CMakeLists.txt [hunk 4]
```diff
@@ -253,6 +253,9 @@ if(ENABLE_TESTING)
     GR_ADD_CPP_TEST("blocks_${qa_file}"
       ${CMAKE_CURRENT_SOURCE_DIR}/${qa_file}
     )
+    set_tests_properties("blocks_${qa_file}"
+        PROPERTIES TIMEOUT 30
+    )
```
Comment (argilo) [inline gr-blocks/lib/CMakeLists.txt:258]: I don't think this change needs to be a part of this PR.

File: gnuradio-runtime/lib/hier_block2_detail.cc [hunk 5]
```diff
@@ -749,11 +754,8 @@ void hier_block2_detail::flatten_aux(flat_flowgraph_sptr sfg) const
     }
     */
 
-    // Construct unique list of blocks used either in edges, inputs,
-    // outputs, or by themselves.  I still hate STL.
```
Comment (argilo) [inline gnuradio-runtime/lib/hier_block2_detail.cc:753]: This change, and the line deletion below, look unrelated to this PR.

**Comment by argilo** [comment]
I removed all the workarounds from Gqrx on this branch: https://github.com/gqrx-sdr/gqrx/compare/remove-unlock-workarounds

So far I haven't been able to trigger crashes/freezes in the scenarios described in https://github.com/gqrx-sdr/gqrx/issues/1030#issuecomment-1007881038.

I'd like to continue with a bit more testing & code review tomorrow before signing off on this.

**Comment by argilo** [comment]
It might be nice to have a test case to exercise the "multiple parents" code path and verify that it raises the correct exception.

File: gr-blocks/python/blocks/qa_hier_block2.py (hunk 2 above)
Comment (vladisslav2011) [reply gr-blocks/python/blocks/qa_hier_block2.py:496]: The order is correct.
`tb.run()` calls `wait()`, so the flowgraph is not running at the time of `tb.disconnect()`
https://github.com/gnuradio/gnuradio/blob/main/gnuradio-runtime/lib/top_block.cc#L66-L70

File: gr-blocks/lib/qa_gr_hier_block2_derived.cc (hunk 3 above)
Comment (vladisslav2011) [reply gr-blocks/lib/qa_gr_hier_block2_derived.cc:103]: No.
We should disconnect first to trigger the bug, then call `lock()` to get the freeze happen.

File: gr-blocks/lib/CMakeLists.txt (hunk 4 above)
Comment (vladisslav2011) [reply gr-blocks/lib/CMakeLists.txt:258]: This is required for the test to fail.
There is no default timeout, so the test would hang indefinitely.

File: gnuradio-runtime/lib/hier_block2_detail.cc (hunk 5 above)
Comment (vladisslav2011) [reply gnuradio-runtime/lib/hier_block2_detail.cc:753]: I agree. Thanks for suggestion.
I'll force-push updated version soon.

**Comment by vladisslav2011** [comment]
@argilo Thanks. I'll add "multiple parents" test to the next version. Converting to draft until new version gets ready.

File: gr-blocks/python/blocks/qa_hier_block2.py (hunk 2 above)
Comment (argilo) [reply gr-blocks/python/blocks/qa_hier_block2.py:496]: Oh, I was confusing `run()` and `start()`.

I don't think `lock()` does anything for a stopped flowgraph, so it might be a better test case if that was replaced with `start()`.

File: gr-blocks/lib/CMakeLists.txt (hunk 4 above)
Comment (argilo) [reply gr-blocks/lib/CMakeLists.txt:258]: OK, if there's no timeout at all, I suppose it's a good improvement to the test suite.

File: gr-blocks/lib/qa_gr_hier_block2_derived.cc (hunk 3 above)
Comment (argilo) [reply gr-blocks/lib/qa_gr_hier_block2_derived.cc:103]: I guess it depends whether you want to test `lock()` or `unlock()`. I believe they both had the same bug, so it might actually be good to test both.

**Review by argilo — APPROVED** [review]
Looks good, and this has tested well with my Gqrx branch.

I did run into one crash in gr-osmosdr's `rtl_source_c` block, but it occurs even on GNU Radio's main branch, so it's unrelated to this pull request. I'd guess it's a race condition in gr-osmosdr.

**Review by mormj — APPROVED** [review]
